naughtydog-decrypter/main.c

```diff
@@ -18,14 +22,118 @@
 #define SHA1_HMAC_KEY   "xM;6X%/p^L/:}-5QoA+K8:F*M!~sb(WK<E%6sW_un0a[7Gm6,()kHoXY+yI/s;Ba"
 
 #define CRC32_POLY    0xEDB88320
+#define CRC32C_POLY   0x82F63B78
 #define CRC32_INIT    0xFFFFFFFF
 
+typedef enum {
+	// PS3: The Last of Us, Uncharted 2, Uncharted 3
+	// PS4: The Last of Us, Uncharted 2 Remastered, Uncharted 3 Remastered, Uncharted 4, The Lost Legacy
+	A = 'A',
+	// PS4: The Last of Us Part 2
+	B = 'B',
+	// PS4: The Nathan Drake Collection
+	C = 'C'
+} SaveType;
+
+typedef struct {
+	u8 *data;
+	u8 start_off;
+	u32 dsize;
+	bool isPS3;
+	SaveType type;
+} SaveState;
+
+void init_savestate(SaveState *ss, u8 *data, size_t len, char type)
+{
+	assert(type == A || type == B || type == C);
+	ss->type = type;
+	switch (ss->type)
+	{
+		case A:
+			ss->start_off = 0x8;
+			break;
+		case B:
+			ss->start_off = 0x10;
+			break;
+		case C:
+			ss->start_off = 0xC;
+			break;
+	}
+
+	if (len < 0xC)
+	{
+		printf("[!] Save is too small!\n");
+		ss->data = NULL;
+		return;
+	}
+	// Check for PS3 saves (big-endian)
+	ss->isPS3 = (data[0] == 0 && data[1] == 0);
+
+	u32 dsize;
+	if (ss->type == B)
+		dsize = *(u32 *)&data[0x8];
+	else
+	{
+		dsize = *(u32 *)&data[len-4];
+		if (ss->isPS3)
+			dsize = ES32(dsize);
+	}
+	if (ss->start_off + dsize > len || !dsize)
+	{
+		printf("[!] Invalid dsize!\n");
+		ss->data = NULL;
+		return;
+	}
+	ss->dsize = dsize;
+	ss->data = data;
+	return;
+}
+
+void swap_u32_data(u8 *data, u32 size)
+{
+	size &= 0xFFFFFFFC;
+	u32 count = size / 4;
+	if (count == 0)
+		return;
```

Comment:
we can skip the count=0 check, the `for` loop will skip anyway 

```suggestion
	u32 count = (size & 0xFFFFFFFC) / 4;
```